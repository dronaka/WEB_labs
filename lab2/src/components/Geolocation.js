import React from "react";
import { connect } from "react-redux";

import { setGeolocation, fetchWeatherByCoords, fetchGeoError } from "../actions/geoActions";
import City from "./City";

import "./styles/Geolocation.css";


class Geolocation extends React.Component {
  componentDidMount() {
    this.getGeolocation();
  }

  render() {
    return (
      <div className="geolocation">
        <h1 className="header">Refresh geolocation</h1>
        <button className="button"
          onClick={() => this.handleClick()}
        >Get geolocation</button>
        {!this.props.error ? this.props.coords && (
          <City
            onFetch={() => this.props.fetchWeatherByCoords(this.props.coords)}
            forecast={this.props.forecast}/>
        ) : (
          <div className="error">Error: {this.props.error}</div>
        )}
      </div>
    );
  }

  handleClick() {
    this.getGeolocation();
  }

  getGeolocation() {
    if (navigator.geolocation) {
      navigator.geolocation.getCurrentPosition(position => {
        const coords = {
          lat: position.coords.latitude,
          lon: position.coords.longitude
        };
        this.props.setGeolocation(coords);
        this.props.fetchWeatherByCoords(this.props.coords);
      },
      () => {
        this.props.setGeolocation({lat: 46.47, lon: 41.53});
        this.props.fetchWeatherByCoords(this.props.coords);
      });
    } else {
      this.props.fetchGeoError("your browser does not support geolocation");
    } 
  }
}


function mapStateToProps(state) {
  return {
    coords: state.geo.coords,
    forecast: state.geo.forecast,
    error: state.geo.error
  };
}

function mapDispatchToProps(dispatch) {
  return {
    setGeolocation: (coords) => {
      dispatch(setGeolocation(coords));
    },

    fetchWeatherByCoords: (coords) => {
      dispatch(fetchWeatherByCoords(coords));
    },

    fetchGeoError: (error) => {
      dispatch(fetchGeoError(error));
    }
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Geolocation);