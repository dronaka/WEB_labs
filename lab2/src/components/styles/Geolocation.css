.geolocation {
    display: flexbox;
    justify-content: space-around;
  }
  
  .geolocation > .header {
    display: inline-block;
  }
  
  .geolocation > .button {
    margin-left: 60px;
    height: 3em;
  }