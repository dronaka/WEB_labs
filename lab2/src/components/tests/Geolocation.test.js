import React from "react";
import thunk from "redux-thunk";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";

import Geolocation from "../Geolocation";


describe("<Geolocation /> render", () => {

  const DEFAULT_FORECAST = {
    cityName: "New York",
    temperature: "-22",
    icon: "icon_code",
    windSpeed: 15,
    description: "good to stay at home",
    pressure: 234,
    humidity: 1,
    coords: {
      lat: 74,
      lon: 89
    }
  };

  const DEFAULT_COORDS = {
    lat: 12,
    lon: 34
  };

  const storeCreator = configureStore([thunk]);

  beforeAll(() => {
    global.fetch = jest.fn(() => Promise.resolve({
      ok: true,
      json: () => Promise.resolve("fake success response")}
      ))
  });

  afterAll(() => {
    delete global.fetch;
  });

  test("should has forecast", () => {
    const store = storeCreator({
      geo: {
        coords: DEFAULT_COORDS,
        forecast: DEFAULT_FORECAST
      }
    });

    const tree = renderer
      .create(
        <Provider store={store}>
          <Geolocation />
        </Provider>)
      .toJSON();

      expect(tree).toMatchSnapshot();
  });

  test("should has error", () => {
    const store = storeCreator({
      geo: {
        error: "test error message"
      }
    });

    const tree = renderer
      .create(
        <Provider store={store}>
          <Geolocation />
        </Provider>
      )
      .toJSON();

      expect(tree).toMatchSnapshot();
  });

});