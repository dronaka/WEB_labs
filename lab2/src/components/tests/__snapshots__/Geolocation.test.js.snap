// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Geolocation /> render should has error 1`] = `
<div
  className="geolocation"
>
  <h1
    className="header"
  >
    Refresh geolocation
  </h1>
  <button
    className="button"
    onClick={[Function]}
  >
    Get geolocation
  </button>
  <div
    className="error"
  >
    Error: 
    test error message
  </div>
</div>
`;

exports[`<Geolocation /> render should has forecast 1`] = `
<div
  className="geolocation"
>
  <h1
    className="header"
  >
    Refresh geolocation
  </h1>
  <button
    className="button"
    onClick={[Function]}
  >
    Get geolocation
  </button>
  <div
    className="city"
  >
    <div
      className="header"
    >
      <div
        className="city-name"
      >
        New York
      </div>
      <div
        className="temperature"
      >
        -22
         â„ƒ
      </div>
      <img
        alt="Weather icon"
        src="https://openweathermap.org/img/wn/icon_code.png"
      />
    </div>
    <div
      className="param"
    >
      <div
        className="name"
      >
        Wind
      </div>
      <div
        className="value"
      >
        15 m/s
      </div>
    </div>
    <div
      className="param"
    >
      <div
        className="name"
      >
        Cloudness
      </div>
      <div
        className="value"
      >
        good to stay at home
      </div>
    </div>
    <div
      className="param"
    >
      <div
        className="name"
      >
        Pressure
      </div>
      <div
        className="value"
      >
        234 hPa
      </div>
    </div>
    <div
      className="param"
    >
      <div
        className="name"
      >
        Humidity
      </div>
      <div
        className="value"
      >
        1%
      </div>
    </div>
    <div
      className="param"
    >
      <div
        className="name"
      >
        Coords
      </div>
      <div
        className="value"
      >
        74, 89
      </div>
    </div>
  </div>
</div>
`;
